packages/identity/src/idp-identity.ts

```diff
@@ -0,0 +1,339 @@
+  readonly cognitoDomain?: CognitoDomainOptions;
+  readonly customDomain?: CustomDomainOptions;
+}
+
+/**
+ * Creates a user pool will one or more federated identity providers
+ */
+export class IdpIdentity extends Construct {
+  public userPool: IUserPool;
+  public userPoolClient?: UserPoolClient;
+  protected hasIdentityPool: boolean = false;
+  protected identityProviders: Array<IUserPoolIdentityProvider> = [];
+
+  constructor(scope: Construct, id: string, readonly props?: IdpIdentityProps) {
+    super(scope, id);
+    Annotations.of(this).addInfo(`${JSON.stringify(props)}`);
+
+    if (props?.userPool) {
+      this.userPool = props.userPool!;
+    } else {
+      this.userPool = this._createUserPool(id, props);
+      this._addDomain(id, this.userPool, {
+        cognitoDomain: props?.cognitoDomain,
+        customDomain: props?.customDomain,
+      });
+      this._attachIdentityProviders(
+        id,
+        props?.identityProviders!,
+        this.userPool
+      );
+    }
+  }
+
+  public addClientApplication(
+    id: string,
+    props?: AddClientProps
+  ): UserPoolClient {
+    const client = new UserPoolClient(this, id, {
+      userPoolClientName: id,
+      userPool: this.userPool,
+      oAuth: {
+        callbackUrls: props?.callbackUrls ? props.callbackUrls : [],
+        logoutUrls: props?.logoutUrls ? props.logoutUrls : [],
+      },
+      authFlows: {
+        userPassword: true,
+        userSrp: true,
+      },
+      generateSecret: true,
+      supportedIdentityProviders: props?.useIdentityProvider,
+      ...props,
+    });
+
+    client.node.addDependency(this.userPool);
+
+    !this.hasIdentityPool &&
+      this._createIdentityPool(id, client).node.addDependency(client);
+
+    return client;
+  }
+
+  /** @internal */
+  protected _addDomain(
+    id: string,
+    userPool: IUserPool,
+    props: CreateUserPoolProps
+  ): void {
+    if (props?.cognitoDomain) {
+      Annotations.of(this).addInfo("cognitoDomain");
+      new UserPoolDomain(this, `${id}-cognitoDomain`, {
+        cognitoDomain: props?.cognitoDomain,
+        userPool: userPool,
+      });
+    } else if (props?.customDomain) {
+      Annotations.of(this).addInfo("customDomain");
+      new UserPoolDomain(this, `${id}-customDomain`, {
+        customDomain: props?.customDomain,
+        userPool: userPool,
+      });
+    }
+  }
+
+  /** @internal */
+  protected _createUserPool(id: string, props?: CreateUserPoolProps): UserPool {
+    const ret = new UserPool(this, "UserPool", {
+      userPoolName: id,
+      removalPolicy: RemovalPolicy.RETAIN,
+      deletionProtection: true,
+      passwordPolicy: defaultPasswordPolicy,
+      mfa: Mfa.REQUIRED,
+      accountRecovery: AccountRecovery.EMAIL_ONLY,
+      autoVerify: {
+        email: true,
+      },
+      ...props,
+    });
+
+    (ret.node.defaultChild as CfnUserPool).userPoolAddOns = {
+      advancedSecurityMode: "ENFORCED",
+    };
+
+    return ret;
```

Comment:
Perhaps we could refactor this into an internal construct that's shared by this and `UserIdentity`?

Reply:
would require a refactor of user-identity/UserIdentity - could be done by making a prop for creating the identity pool. not sure of we want to change UserIdentity though

Reply:
I think it'd just be a construct like:

```ts
class UserPoolWithDefaults extends UserPool {
  constructor(scope: Construct, id: string, props: UserPoolProps) {
    super(scope, id, {
      removalPolicy: RemovalPolicy.RETAIN,
      deletionProtection: true,
      passwordPolicy: defaultPasswordPolicy,
      mfa: Mfa.REQUIRED,
      accountRecovery: AccountRecovery.EMAIL_ONLY,
      autoVerify: {
        email: true,
      },
    });
    (this.node.defaultChild as CfnUserPool).userPoolAddOns = {
      advancedSecurityMode: "ENFORCED",
    };
  }
}
```

Then make use of it in both `UserIdentity` and `IdpIdentity` :)